litmus/libdir/_aarch64/kvm_fault_handler.c

```diff
@@ -9,13 +9,22 @@ static void record_fault(who_t *w, unsigned long pc, unsigned long esr) {
   fault_info_t flt;
   unsigned long far;
 
+  flt.type = get_fault_type(esr);
+
+  if(flt.type == FaultSupervisorCall) {
+    // SVC, SMC and HVC return address points to the next instruction
+    // after the exception generating instruction. For the purposes
+    // of fault recording, we are interested in the instruction that
+    // generated the fault.
```

Comment:
Thanks, @murzinv, it's an interesting point. I agree that it is a matter of convention whether in `Fault(P0:L0,...)` L0 refers to the instruction that caused the fault or to the content of ELR. And indeed it seems that so far one could pretend that P0:L0 is the content of ELR, so the convention is currently not clear.

This interpretation would be a bit of a stretch in litmus tests with fault handlers, where we commonly modify ELR inside of the fault handler to return to the next line: seems like possibility of such changes would make the `Fault` predicate difficult to interpret if it referred to ELR. Additionally, this label argument is optional. Overall, my hunch is that the convention of letting that label (when provided) refer to the instruction that caused the fault would be easier to use, but it's not a strong opinion.

Whichever convention we use, it's worth ensuring it's documented, although maybe that could be a different PR.

Reply:
That's a very good point @murzinv.

> his interpretation would be a bit of a stretch in litmus tests with fault handlers, where we commonly modify ELR inside of the fault handler to return to the next line: seems like possibility of such changes would make the Fault predicate difficult to interpret if it referred to ELR. Additionally, this label argument is optional. Overall, my hunch is that the convention of letting that label (when provided) refer to the instruction that caused the fault would be easier to use, but it's not a strong opinion.

@artkhyzha I think faults atoms are a bit special in that regard. The `fault()` atoms are the only atoms that don't ask a question about the final state. Take this test for example:

```
AArch64 LDRv0+I2V
Variant=vmsa,handled
{
 [PTE(x)]=(oa:PA(x),valid:0);

 0:X1=x;
 0:X3=PTE(x);
 0:X2=(oa:PA(x),valid:1);
}
 P0          | P1          ;
L0:          |             ;
 LDR W0,[X1] | STR X2,[X3] ;
exists(fault(P0:L0,x))
```
For this tests, there is an execution where the LDR is executed multiple times, generating a fault and then one more last time that doesn't generate a fault. In this execution the post-condition is satisfied, despite that the last execution of LDR didn't fault. So strictly speaking the fault atom is not an atom about the final state.

Having said that, I don't have a strong preference, either way seems fine to me because we can easily explain it:
1. <label> is the value of the `PC` register when it was determined that we need to take an exception.
2. <label> is the value of the `ELR_EL1` register when we took the exception.

It looks like currently litmus7 is recording the `PC` and not the `ELR_EL1`.

Reply:
Sorry,  I just realised that this statement:

> Having said that, I don't have a strong preference, either way seems fine to me because we can easily explain it:
> 1.  <label> is the value of the PC register when it was determined that we need to take an exception.
> 2. <label> is the value of the ELR_EL1 register when we took the exception.

I just had a look at code of this PR and it seems that @diaolo01 has to adjust the value in the `PC`. Based on that my vote would be to not make the adjustment and define that <label> is the value of the `PC` when it is determined that an exception needs to be taken.